### PR TITLE
TOK-412: List all activated builders

### PR DESCRIPTION
**Pull request by DinizSa**
What
- List all the activated Builders

[Jira Ticket](https://rsklabs.atlassian.net/jira/software/projects/TOK/boards/267/backlog?assignee=712020%3A42dabdd3-0b32-4060-be21-62cfd46a1751&selectedIssue=TOK-412)

### CODE REVIEW

File: src/app/collective-rewards/user/hooks/useGetBuilders.ts [hunk 1]
```diff
@@ -124,6 +133,7 @@ export const useGetBuilders = (): BuildersLoader => {
           ? builderStatusMap[builder as Address] // V2
           : BuilderStatusProposalCreatedMVP, // MVP
       proposals: Object.values(proposals),
+      gauge: builderToGauge?.[builder as Address],
```
Comment (franciscotobar) [inline src/app/collective-rewards/user/hooks/useGetBuilders.ts:136]: @jurajpiar here.

File: src/app/collective-rewards/leaderboard/BuildersLeaderBoardTable.tsx [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+    direction: 'asc',
+  })
+
+  // pagination
+  const buildersPerPage = 10
+  const tableDataLength = useMemo(() => Object.keys(rewardsData).length, [rewardsData])
+  const maxPages = useMemo(() => Math.ceil(tableDataLength / buildersPerPage), [tableDataLength])
+  const { paginationUi, currentPage } = useBasicPaginationUi(maxPages)
+
+  useEffect(() => {
```
Comment (franciscotobar) [inline src/app/collective-rewards/leaderboard/BuildersLeaderBoardTable.tsx:234]: Here we already have a method to [handle](https://github.com/RootstockCollective/dao-frontend/blob/develop/src/app/collective-rewards/utils/handleErrors.ts) errors.

File: src/app/collective-rewards/leaderboard/hooks/useGetBuildersRewards.ts [hunk 3]
```diff
@@ -1,71 +1,230 @@
...
+    error: rewardsERC20Error,
+  } = useGetRewardsERC20()
+
+  const {
+    data: rewardsCoinbase,
+    isLoading: rewardsCoinbaseLoading,
+    error: rewardsCoinbaseError,
+  } = useGetRewardsCoinbase()
+
+  const { cycleStartTimestamp, cycleEndTimestamp } = getPreviousCycle()
```
Comment (franciscotobar) [inline src/app/collective-rewards/leaderboard/hooks/useGetBuildersRewards.ts:55]: There is a mix on this, this method is depecrated and needs to be removed, the information from the epochs its under the CycleContext.

File: src/app/collective-rewards/leaderboard/hooks/useGetBuildersRewards.ts
Comment (antomor) [inline src/app/collective-rewards/leaderboard/hooks/useGetBuildersRewards.ts:1]: @DinizSa I guess we should remove the mocked data from here

File: src/app/collective-rewards/rewards/hooks/getTotalPotentialRewards.ts [hunk 4]
```diff
@@ -14,7 +14,7 @@ export const useGetTotalPotentialReward = () => {
   })
 
   return {
-    data,
+    data: data || 0n,
```
Comment (franciscotobar) [inline src/app/collective-rewards/rewards/hooks/getTotalPotentialRewards.ts:17]: We usually handle the undefined from where it's being called.  Right now if you check the other hooks, we pass the data as it is.

File: src/app/collective-rewards/leaderboard/hooks/useGetBuildersRewards.ts
Comment (DinizSa) [reply src/app/collective-rewards/leaderboard/hooks/useGetBuildersRewards.ts:1]: yes, done

File: src/app/collective-rewards/user/context/BuilderContext.tsx [hunk 5]
```diff
@@ -77,6 +78,7 @@ export const BuilderContextProvider: FC<BuilderProviderProps> = ({ children }) =
           proposalName,
           proposalDescription,
           joiningDate,
+          gauge: builder.gauge,
```
Comment (franciscotobar) [inline src/app/collective-rewards/user/context/BuilderContext.tsx:81]: I would destruct the builder object in line 62. 

` const { status, address, gauge } = builder`

And just pass the property:
` gauge`

**Comment by franciscotobar** [comment]
I allocated some rewards and we need to update the formating:
Screenshot 2024-11-19 at 8 09 14 AM

File: src/app/collective-rewards/leaderboard/hooks/useGetBuildersRewards.ts [hunk 6]
```diff
@@ -1,68 +1,144 @@
...
+
   return {
-    data: whitelistedBuilders.map(({ address, builderName }) => {
-      const builderEvents = rewardDistributedLogs.filter(event =>
-        isAddressEqual(event.args.builder_, address),
-      )
-      const lastCycleRewards = getLastCycleRewards(builderEvents)
-      const lastCycleRewardsInHuman = Number(formatBalanceToHuman(lastCycleRewards))
+    data: whitelistedBuilders.map(({ address, builderName }, i) => {
+      const projectedRif = rewardShares && rewardsERC20 ? rewardShares[i] * rewardsERC20 : 0n
```
Comment (franciscotobar) [inline src/app/collective-rewards/leaderboard/hooks/useGetBuildersRewards.ts:95]: We have the shares and the rewards but we also need to take into account the backer reward %

File: src/app/collective-rewards/leaderboard/BuildersLeaderBoardTable.tsx [hunk 7]
```diff
@@ -0,0 +1,423 @@
...
+  return (
+    <TableCell className={cn(tableHeaders[5].className, 'border-solid align-center')}>
+      <Button variant="secondary" /* disabled={true} */>Select</Button>
+    </TableCell>
+  )
+}
+
+enum RewardsColumnKeyEnum {
+  builder = 'builder',
+  lastCycleReward = 'lastCycleReward',
```
Comment (franciscotobar) [inline src/app/collective-rewards/leaderboard/BuildersLeaderBoardTable.tsx:195]: I would try to keep it in plura.
```suggestion
  lastCycleRewards = 'lastCycleRewards',
```

File: src/app/collective-rewards/leaderboard/BuildersLeaderBoardTable.tsx [hunk 8]
```diff
@@ -0,0 +1,423 @@
...
+  projectedReward = 'projectedReward',
```
Comment (franciscotobar) [inline src/app/collective-rewards/leaderboard/BuildersLeaderBoardTable.tsx:196]: I would keep the same name, 
```suggestion
  estimatedRewards = 'estimatedRewards',
```

File: src/app/collective-rewards/leaderboard/BuildersLeaderBoardTable.tsx [hunk 9]
```diff
@@ -0,0 +1,423 @@
...
+  { label: 'Builder', className: 'w-[14%]', key: RewardsColumnKeyEnum.builder },
+  { label: 'Backer Rewards %', className: 'w-[10%]', key: RewardsColumnKeyEnum.rewardPercentage },
+  { label: 'Last Cycle Rewards', className: 'w-[22%]', key: RewardsColumnKeyEnum.lastCycleReward },
+  { label: 'Est. Backers Rewards', className: 'w-[22%]', key: RewardsColumnKeyEnum.projectedReward },
+  {
+    label: 'Total Allocations',
+    className: 'w-[18%]',
+    // eslint-disable-next-line quotes
+    tooltip: "The Builder's share of the total allocations",
+    key: RewardsColumnKeyEnum.share,
```
Comment (franciscotobar) [inline src/app/collective-rewards/leaderboard/BuildersLeaderBoardTable.tsx:210]: Why do you use share? Isnt totalAllocation?

**Comment by franciscotobar** [comment]
I may be wrong, but we are still missing the flags in the builder name.
Screenshot 2024-11-19 at 10 11 34 AM

File: src/components/ProgressBar.tsx [hunk 10]
```diff
@@ -0,0 +1,19 @@
+import React from 'react'
+
+type Props = {
```
Comment (franciscotobar) [inline src/components/ProgressBar.tsx:3]: This needs to be inside a folder to keep the same structure

File: src/app/collective-rewards/user/hooks/useGetBuilders.ts [hunk 11]
```diff
@@ -83,6 +84,14 @@ export const useGetBuilders = (): BuildersLoader => {
   })
   const builders = buildersResult?.map(builder => builder.result) as Address[]
 
+  let builderToGauge = builders?.reduce(
```
Comment (franciscotobar) [inline src/app/collective-rewards/user/hooks/useGetBuilders.ts:87]:
```suggestion
  const builderToGauge = builders?.reduce(
```

File: src/app/collective-rewards/leaderboard/BuildersLeaderBoardTable.tsx [hunk 12]
```diff
@@ -0,0 +1,423 @@
...
+export const RewardCell: FC<RewardCellProps> = ({ rewards }) => (
+  <div className="flex flex-nowrap flex-row gap-1">
+    {rewards &&
+      rewards.map(({ crypto: { value, symbol }, fiat: { value: fiatValue, symbol: fiatSymbol } }) => (
+        <div key={value + symbol} className="flex-1">
+          {/* TODO: if the value is very small, should we show it in Gwei/wei? */}
+
+          <Label className="flex items-center gap-1 font-normal text-sm leading-none text-text-primary font-rootstock-sans">
+            {toFixed(value)}
+            {symbol === RBTC && (
```
[the remark this answers is not in the document]
Comment (DinizSa) [reply src/app/collective-rewards/leaderboard/BuildersLeaderBoardTable.tsx:54]: RewardCell receives rewards, and is not called individually, do you have any suggestion on how to receive the icon as input?
Meanwhile I created a separate component `rbtcIcon`, and `rifIcon` to allow reusability

[the remark this answers is not in the document]
Comment (franciscotobar) [reply src/app/collective-rewards/leaderboard/BuildersLeaderBoardTable.tsx:54]: I already addressed it in my PR while doing the abstraction, so if you want let's keep it as it is right now :)

**Comment by franciscotobar** [comment]
> I may be wrong, but we are still missing the flags in the builder name. Screenshot 2024-11-19 at 10 11 34 AM

We agree to handle this in a different task

**Review by franciscotobar — APPROVED** [review]
LGTM

**Pull request merged**